astropy/units/quantity.py

```diff
@@ -382,14 +383,23 @@ def __new__(cls, value, unit=None, dtype=None, copy=True, order=None,
                               or value.dtype.kind == 'O'):
             value = value.astype(float)
 
-        value = value.view(cls)
-        value._set_unit(value_unit)
+        # cast the data to 'cls' (some Quantity subclass)
+        if copy:  # create new array and copy data
+            qvalue = super(Quantity, cls).__new__(
```

Comment:
I realize I suggested it myself, but seeing this, I feel this is rather fragile in case someone creates, say, a `QuantityVariable(Quantity, Variable)` or `MaskedQuantity(Quantity, MaskedArray)` class (I have done the latter, and #3715 only happens to be the reverse of the former)...

If we go this route, it might be safer to actually take the view and then return `value.copy()` (which does own its data).

Reply:
p.s. I'm not sure it was such a good design choice to have the `ndarray` constructor being so different from `np.array`...